### PR TITLE
Add tests with RSpec for the models

### PR DESCRIPTION
**Pull request by nat-sch**
Add RSpec; add tests for the models; add Rubocop

### CODE REVIEW

File: spec/models/initiative_spec.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+#
+# Foreign Keys
+#
+#  parent_id  (parent_id => initiatives.id)
+#
+require 'rails_helper'
+
+RSpec.describe Initiative, type: :model do
+  context 'in its entities,' do
+      it 'may have helpers' do
```
Comment (jicc96) [inline spec/models/initiative_spec.rb]: Some inconsistent tabs here. this should be indented with 4 spaces

File: spec/models/initiative_spec.rb [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+#  index_initiatives_on_parent_id  (parent_id)
+#
+# Foreign Keys
+#
+#  parent_id  (parent_id => initiatives.id)
+#
+require 'rails_helper'
+
+RSpec.describe Initiative, type: :model do
+  context 'in its entities,' do
```
Comment (jicc96) [inline spec/models/initiative_spec.rb]: I encourage you to follow the https://www.betterspecs.org/ nameing conventions for Rspec test

File: spec/models/initiative_spec.rb [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      it 'may have helpers' do
+        source = Wyeworker.new(name: "Source Pablo")
+        initiative_no_helpers = Initiative.new(title: "Test Inititative No Helpers", source: source)
```
Comment (jicc96) [inline spec/models/initiative_spec.rb]: Highly recommend the usage of factory bot

File: spec/models/initiative_spec.rb (hunk 1 above)
Comment (jicc96) [reply spec/models/initiative_spec.rb]: As a more relevant comment you should never tests may things. That is not very specific and usually is a poor behavior descriptor. Saying that I "may" have something only means that I can have it or can not have it so it shouldn't be a test because the code has nothing to enforce. If the object is present is cool if the object is absent then is also cool

File: spec/models/initiative_spec.rb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+      end
+  end
+  
+  context 'when looking back from its entities,' do
+    it 'is an initiative of its source' do
+      source = Wyeworker.new(name: "Source Pablo")
+      initiative = Initiative.create(title: "Test Inititative", source: source)
+
+      expect(source.initiatives).to include(initiative)
+    end
```
Comment (jicc96) [inline spec/models/initiative_spec.rb]: As a rule of thumb if you are not using the entity that you are testing (in this case the initiative) in the left operand of an expectation then something is wrong. This test belongs to `Source` since it's testing it's ability to retrive `Initiciatives`. Also this is something that should probably not be tested as all this connections are generated by the framework (there exists 1 justification that would make testing this reasonable it is left to the reader to present that argument)

File: spec/models/wyeworker_spec.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+require 'rails_helper'
+
+RSpec.describe Wyeworker, type: :model do
+  context 'in its initiatives' do
+    it 'can correctly take up the role of a source' do
+      source = Wyeworker.new(name: "Source Pablo")
+      initiative = Initiative.create(title: "Test Inititative", source: source)
+
+      expect(source.wyeworker_initiative_belongings.first.kind).to eql("source")
+      expect(initiative.source).to be(source)
```
Comment (jicc96) [inline spec/models/wyeworker_spec.rb]: Same as above

File: spec/models/wyeworker_spec.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    end
+
+    it 'can correctly take up the role of a helper' do
+      source = Wyeworker.new(name: "Source Pablo")
+      helper = Wyeworker.new(name: "Helper Pedro", )
+      initiative = Initiative.create(title: "Test Inititative", source: source, helpers: [helper])
+
+      expect(helper.wyeworker_initiative_belongings.first.kind).to eql("helper")
+      expect(initiative.helpers.first).to be(helper)
+    end
```
Comment (jicc96) [inline spec/models/wyeworker_spec.rb]: I'm having some trouble understanding this test. This will read `Wyeworker in its initiatives can correctly take up the role of a helper` but the test uses a helper and a source what are we testing here exactly?

**Review by jicc96 — COMMENTED** [review]
Some general comments:

- Please introduce factory bot to the tests I can explain what it is or what it does but it will help a lot with the code.
- I recommend using `let` statements and `subject` statements to this tests.
- As they stand right now I'm not sure if the models have code worth testing. I suggest adding custom code to them and the usage of callbacks

File: spec/models/initiative_spec.rb (hunk 3 above)
Comment (nat-sch) [reply spec/models/initiative_spec.rb]: Oh, I checked it out, agreed!

File: app/models/initiative.rb [hunk 7]
```diff
@@ -49,15 +49,15 @@ class Initiative < ApplicationRecord
   validate :must_have_manager
 
   def must_have_manager
-    if not (
-        source.is_a?(Manager) or
-        helpers.any? { |h| h.is_a?(Manager) }
-      )
-        errors.add :wyeworker_initiative_belongings, "An initiative must have a manager involved, as a source or as a helper"
+    unless source.is_a?(Manager) ||
```
Comment (jdelmazo9) [inline app/models/initiative.rb]: It's not recommended to use unless with multiple conditions since it can be hard to follow. Anyways an option to keep using unless could be using an auxiliar method.
It could be something like
`unless manager_is_involved`

```
def manager_is_involved
  source.is_a?(Manager) || helpers.any? { |h| h.is_a?(Manager) }
end
```

File: spec/models/wyeworker_spec.rb [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+#  created_at :datetime         not null
+#  updated_at :datetime         not null
+#
+require "rails_helper"
+
+RSpec.describe Wyeworker, type: :model do
+  it "wires up properly as a source" do
+    source = create(:wyeworker_source)
+
+    expect(source.wyeworker_initiative_belongings.first.kind).to eql("source")
```
Comment (jdelmazo9) [inline spec/models/wyeworker_spec.rb]: I feel this is testing more the things that you did on factorybot than logic related to your program, might not be necessary 🤔 Anyways if it would be needed I think I'd write the test but with all the steps here.
I mean, something like:
- create a wyeworker
- create initiative
- link wyeworker as initiative's source
- validate

File: spec/models/initiative_spec.rb [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+RSpec.describe Initiative, type: :model do
+  let(:initiative) { create(:initiative) }
+
+  it "wires up as an initiative of its source" do
+    expect(initiative.source.initiatives).to include(initiative)
+  end
+
+  it "wires up as an initiative of its helpers" do
+    expect(initiative.helpers.first.initiatives).to include(initiative)
```
Comment (jdelmazo9) [inline spec/models/initiative_spec.rb]: if you want to do something more dynamic you could use `.sample` instead of `.first`, or just validate that all helpers initiatives include the initiative.

File: app/models/initiative.rb (hunk 7 above)
Comment (nat-sch) [reply app/models/initiative.rb]: Got it! Rubocop by default preferred `unless`, but I configured it! So this should be fixed

File: spec/models/wyeworker_spec.rb (hunk 8 above)
Comment (nat-sch) [reply spec/models/wyeworker_spec.rb]: There's lambdas and quite a bit of configuration involved, that's what I wanted to test. I see your point about being explicit, I changed the test accordingly!

File: spec/models/initiative_spec.rb (hunk 9 above)
Comment (nat-sch) [reply spec/models/initiative_spec.rb]: Oh I love that! Done!

**Comment by nat-sch** [comment]
Foks reviewed this and we agreed it looks good
